src/raft.rs

```diff
@@ -884,16 +998,137 @@ impl<T: Storage> Raft<T> {
     pub fn bcast_append(&mut self) {
         let self_id = self.id;
-        let core = &mut self.r;
-        let msgs = &mut self.msgs;
-        self.prs
-            .iter_mut()
-            .filter(|&(id, _)| *id != self_id)
-            .for_each(|(id, pr)| core.send_append(*id, pr, msgs));
+        let mut msgs: Vec<Message> = Vec::default();
+        {
+            // Messages are stored in a vector temporarily.
+            // They will be pushed to message queue later.
+            let core = &mut self.r;
+            self.prs
+                .iter_mut()
+                .filter(|&(id, _)| *id != self_id)
+                .for_each(|(id, pr)| core.send_append(*id, pr, &mut msgs));
+        }
+
+        // Use leader replication if follower replication is disabled or
+        // the broadcast group id of leader is unknown. Broadcast MsgAppend
+        // as normal.
+        let leader_group_id = self
+            .prs()
+            .get(self_id)
+            .map_or(0, |pr| pr.broadcast_group_id);
+        if !self.follower_replication() || leader_group_id != 0 {
+            self.msgs.append(&mut msgs);
+            return;
+        }
+
+        // If follower replication is enabled, MsgAppends sent to the same broadcast group
+        // will be merge into a MsgGroupBroadcast.
+        //
+        // Record message that should be discarded after merging.
+        let mut skip = vec![false; msgs.len()];
+        // Message group:
+        // broadcast group id -> {index of messages in msgs}
+        let mut msg_group: HashMap<u64, Vec<usize>> = HashMap::default();
+
+        // Iterate messages generated by leader.
+        // Filter out messages that should be sent by follower replication,
+        // and group them by broadcast group id.
+        for (pos, msg) in msgs.iter().enumerate() {
+            // Only reserve MsgAppend sent to peers in replicating state
+            if msg.get_msg_type() != MessageType::MsgAppend || !self.is_replicating(msg.to) {
+                continue;
+            }
+            // Get the broadcast group id of target peer.
+            let group_id = self.prs().get(msg.to).map_or(0, |pr| pr.broadcast_group_id);
+            // Do not need merge if broadcast group id is unknown or in the same
+            // group with leader. Valid broadcast group id should be greater than 0.
+            if group_id == 0 || group_id != leader_group_id {
+                continue;
+            }
+
+            // Group messages
+            if let Some(group) = msg_group.get_mut(&group_id) {
+                group.push(pos);
+            } else {
+                msg_group.insert(group_id, vec![pos]);
+            }
+        }
+
+        // Merge MsgAppend in broadcast groups.
+        for (_, group) in msg_group {
+            self.merge_append_group(&group, &mut msgs, &mut skip);
+        }
+
+        let mut idx: usize = 0;
+        for msg in msgs {
+            if !skip[idx] {
```

Comment:
How about grouping the messages instead of the pos. And `merge_append_group` modifies or filters the messages directly instead of by passing `skip`

Reply:
In new commit, I have modified it. PTAL.